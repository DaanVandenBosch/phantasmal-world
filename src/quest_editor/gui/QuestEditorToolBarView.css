.quest_editor_QuestEditorToolBarView_save_as_dialog_content {
    display: grid;
    grid-template-columns: 100px max-content;
    align-items: center;
}
