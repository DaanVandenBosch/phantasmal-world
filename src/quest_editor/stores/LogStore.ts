import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { Property } from "../../core/observable/property/Property";
import { list_property, property } from "../../core/observable";
import { Disposable } from "../../core/observable/Disposable";
import { Disposer } from "../../core/observable/Disposer";
import Logger from "js-logger";
import { enum_values } from "../../core/enums";

export type QuestLogger = {
    readonly debug: (message: string) => void;
    readonly info: (message: string) => void;
    readonly warning: (message: string) => void;
    readonly error: (message: string) => void;
};

// Log level names in order of importance.
export enum LogLevel {
    Debug,
    Info,
    Warning,
    Error,
}

export const LogLevels = enum_values<LogLevel>(LogLevel);

export type LogMessage = {
    readonly formatted_timestamp: string;
    readonly message: string;
    readonly level: LogLevel;
};

export class LogStore implements Disposable {
    private readonly disposer = new Disposer();
    private readonly default_log_level = LogLevel.Info;
    private readonly _log_messages = list_property<LogMessage>();
    private readonly _log_level = property<LogLevel>(this.default_log_level);

    readonly log_messages: ListProperty<LogMessage>;
    readonly log_level: Property<LogLevel> = this._log_level;

    constructor() {
        this.log_messages = this._log_messages.filtered(
            this.log_level.map(level => message => message.level >= level),
        );
    }

    dispose(): void {
        this.disposer.dispose();
    }

    set_log_level(log_level: LogLevel): void {
        this._log_level.val = log_level;
    }

    private add_log_message(message: string, level: LogLevel, name: string): void {
        this._log_messages.push({
            formatted_timestamp: new Date().toISOString(),
            message,
            level,
        });

        switch (level) {
            case LogLevel.Debug:
                Logger.get(name).debug(message);
                break;
            case LogLevel.Info:
                Logger.get(name).info(message);
                break;
            case LogLevel.Warning:
                Logger.get(name).warn(message);
                break;
            case LogLevel.Error:
                Logger.get(name).error(message);
                break;
        }
    }

    /**
     * @param name - js-logger logger name
     */
    get_logger(name: string): QuestLogger {
        return {
            debug: (message: string): void => {
                this.add_log_message(message, LogLevel.Debug, name);
            },
            info: (message: string): void => {
                this.add_log_message(message, LogLevel.Info, name);
            },
            warning: (message: string): void => {
                this.add_log_message(message, LogLevel.Warning, name);
            },
            error: (message: string): void => {
                this.add_log_message(message, LogLevel.Error, name);
            },
        };
    }
}

export const log_store = new LogStore();
