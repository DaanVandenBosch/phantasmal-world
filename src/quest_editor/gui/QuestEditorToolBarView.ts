import { ToolBar } from "../../core/gui/ToolBar";
import { FileButton } from "../../core/gui/FileButton";
import { Button } from "../../core/gui/Button";
import { undo_manager } from "../../core/undo/UndoManager";
import { Select } from "../../core/gui/Select";
import { div, Icon } from "../../core/gui/dom";
import { DropDown } from "../../core/gui/DropDown";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import {
    AreaAndLabel,
    QuestEditorToolBarController,
} from "../controllers/QuestEditorToolBarController";
import { View } from "../../core/gui/View";
import { Dialog } from "../../core/gui/Dialog";
import { TextInput } from "../../core/gui/TextInput";
import "./QuestEditorToolBarView.css";
import { Version } from "../../core/data_formats/parsing/quest/Version";

export class QuestEditorToolBarView extends View {
    private readonly toolbar: ToolBar;

    get element(): HTMLElement {
        return this.toolbar.element;
    }

    get height(): number {
        return this.toolbar.height;
    }

    constructor(ctrl: QuestEditorToolBarController) {
        super();

        const new_quest_button = new DropDown({
            text: "New quest",
            icon_left: Icon.NewFile,
            items: [Episode.I],
            to_label: episode => `Episode ${Episode[episode]}`,
        });
        const open_file_button = new FileButton({
            icon_left: Icon.File,
            text: "Open file...",
            accept: ".bin, .dat, .qst",
            multiple: true,
            tooltip: "Open a quest file (Ctrl-O)",
        });
        const save_as_button = new Button({
            text: "Save as...",
            icon_left: Icon.Save,
            tooltip: "Save this quest to new file (Ctrl-Shift-S)",
        });
        const undo_button = new Button({
            text: "Undo",
            icon_left: Icon.Undo,
            tooltip: undo_manager.first_undo.map(
                action =>
                    (action ? `Undo "${action.description}"` : "Nothing to undo") + " (Ctrl-Z)",
            ),
        });
        const redo_button = new Button({
            text: "Redo",
            icon_left: Icon.Redo,
            tooltip: undo_manager.first_redo.map(
                action =>
                    (action ? `Redo "${action.description}"` : "Nothing to redo") +
                    " (Ctrl-Shift-Z)",
            ),
        });
        // TODO: make sure select menu is updated when entity counts change.
        const area_select = new Select<AreaAndLabel>({
            items: ctrl.areas,
            to_label: ({ label }) => label,
        });
        const debug_button = new Button({
            text: "Debug",
            icon_left: Icon.Play,
            tooltip: "Debug the current quest in a virtual machine (F5)",
        });
        const resume_button = new Button({
            text: "Continue",
            icon_left: Icon.SquareArrowRight,
            tooltip: "Resume execution (F6)",
        });
        const step_over_button = new Button({
            text: "Step over",
            icon_left: Icon.LongArrowRight,
            tooltip: "Execute the next line and step over any function calls (F8)",
        });
        const step_in_button = new Button({
            text: "Step into",
            icon_left: Icon.LevelDown,
            tooltip: "Execute the next line and step inside any function calls (F7)",
        });
        const step_out_button = new Button({
            text: "Step out",
            icon_left: Icon.LevelUp,
            tooltip: "Execute until outside of current call frame (Shift-F8)",
        });
        const stop_button = new Button({
            text: "Stop",
            icon_left: Icon.Stop,
            tooltip: "Stop execution (Shift-F5)",
        });

        const children = [
            new_quest_button,
            open_file_button,
            save_as_button,
            undo_button,
            redo_button,
            area_select,
        ];

        if (ctrl.vm_feature_active) {
            children.push(
                debug_button,
                resume_button,
                step_over_button,
                step_in_button,
                step_out_button,
                stop_button,
            );
        }

        this.toolbar = this.disposable(new ToolBar(...children));

        // "Save As" dialog.
        const filename_input = this.disposable(
            new TextInput(ctrl.filename.val, { label: "File name:" }),
        );
        const version_select = this.disposable(
            new Select({
                label: "Version:",
                items: [Version.GC, Version.BB],
                selected: ctrl.version,
                to_label: version => {
                    switch (version) {
                        case Version.DC:
                            return "Dreamcast";
                        case Version.GC:
                            return "GameCube";
                        case Version.PC:
                            return "PC";
                        case Version.BB:
                            return "BlueBurst";
                    }
                },
            }),
        );
        const save_button = this.disposable(new Button({ text: "Save" }));
        const cancel_button = this.disposable(new Button({ text: "Cancel" }));

        const save_as_dialog = this.disposable(
            new Dialog({
                title: "Save As",
                visible: ctrl.save_as_dialog_visible,
                content: div(
                    { className: "quest_editor_QuestEditorToolBarView_save_as_dialog_content" },
                    filename_input.label!.element,
                    filename_input.element,
                    version_select.label!.element,
                    version_select.element,
                ),
                footer: [save_button.element, cancel_button.element],
            }),
        );

        save_as_dialog.element.addEventListener("keydown", evt => {
            if (evt.key === "Enter") {
                ctrl.save_as();
            }
        });

        this.disposables(
            new_quest_button.chosen.observe(({ value: episode }) => ctrl.create_new_quest(episode)),

            open_file_button.files.observe(({ value: files }) => ctrl.parse_files(files)),

            save_as_button.onclick.observe(ctrl.save_as_clicked),
            save_as_button.enabled.bind_to(ctrl.can_save),

            save_as_dialog.ondismiss.observe(ctrl.dismiss_save_as_dialog),

            filename_input.value.bind_to(ctrl.filename),
            filename_input.value.observe(({ value }) => ctrl.set_filename(value)),

            version_select.selected.observe(({ value }) => {
                if (value != undefined) {
                    ctrl.set_version(value);
                }
            }),

            save_button.onclick.observe(ctrl.save_as),
            cancel_button.onclick.observe(ctrl.dismiss_save_as_dialog),

            undo_button.onclick.observe(() => undo_manager.undo()),
            undo_button.enabled.bind_to(ctrl.can_undo),

            redo_button.onclick.observe(() => undo_manager.redo()),
            redo_button.enabled.bind_to(ctrl.can_redo),

            area_select.selected.bind_to(ctrl.current_area),
            area_select.selected.observe(({ value }) => ctrl.set_area(value!)),
            area_select.enabled.bind_to(ctrl.can_select_area),

            debug_button.onclick.observe(ctrl.debug),
            debug_button.enabled.bind_to(ctrl.can_debug),

            resume_button.onclick.observe(ctrl.resume),
            resume_button.enabled.bind_to(ctrl.can_step),

            step_over_button.onclick.observe(ctrl.step_over),
            step_over_button.enabled.bind_to(ctrl.can_step),

            step_in_button.onclick.observe(ctrl.step_in),
            step_in_button.enabled.bind_to(ctrl.can_step),

            step_out_button.onclick.observe(ctrl.step_out),
            step_out_button.enabled.bind_to(ctrl.can_step),

            stop_button.onclick.observe(ctrl.stop),
            stop_button.enabled.bind_to(ctrl.can_stop),
        );

        this.finalize_construction();
    }
}
