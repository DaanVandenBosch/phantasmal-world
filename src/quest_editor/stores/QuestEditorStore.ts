import { property, list_property } from "../../core/observable";
import { QuestModel } from "../model/QuestModel";
import { Property, PropertyChangeEvent } from "../../core/observable/property/Property";
import { read_file } from "../../core/read_file";
import { parse_quest, write_quest_qst } from "../../core/data_formats/parsing/quest";
import { ArrayBufferCursor } from "../../core/data_formats/cursor/ArrayBufferCursor";
import { Endianness } from "../../core/data_formats/Endianness";
import { QuestObjectModel } from "../model/QuestObjectModel";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { AreaModel } from "../model/AreaModel";
import { area_store } from "./AreaStore";
import { SectionModel } from "../model/SectionModel";
import { QuestEntityModel } from "../model/QuestEntityModel";
import { Disposable } from "../../core/observable/Disposable";
import { Disposer } from "../../core/observable/Disposer";
import { gui_store, GuiTool } from "../../core/stores/GuiStore";
import { UndoStack } from "../../core/undo/UndoStack";
import { TranslateEntityAction } from "../actions/TranslateEntityAction";
import { EditShortDescriptionAction } from "../actions/EditShortDescriptionAction";
import { EditLongDescriptionAction } from "../actions/EditLongDescriptionAction";
import { EditNameAction } from "../actions/EditNameAction";
import { EditIdAction } from "../actions/EditIdAction";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import { create_new_quest } from "./quest_creation";
import { CreateEntityAction } from "../actions/CreateEntityAction";
import { RemoveEntityAction } from "../actions/RemoveEntityAction";
import { Euler, Vector3 } from "three";
import { RotateEntityAction } from "../actions/RotateEntityAction";
import { convert_quest_from_model, convert_quest_to_model } from "./model_conversion";
import Logger = require("js-logger");
import { MessageLogStore, LogMessage, LogLevel, LogGroup } from "../../core/gui/MessageLog";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { WritableProperty } from "../../core/observable/property/WritableProperty";
import { QuestRunner } from "../QuestRunner";

const logger = Logger.get("quest_editor/gui/QuestEditorStore");

export interface Logger {
    debug(...items: any[]): void;
    info(...items: any[]): void;
    warning(...items: any[]): void;
    error(...items: any[]): void;
}

export class QuestEditorStore implements Disposable, MessageLogStore {
    private readonly disposer = new Disposer();
    public readonly quest_runner: QuestRunner = new QuestRunner();
    /**
     * Log levels as Record<name, LogLevel>
     */
    private readonly log_levels_map = (function<KT extends readonly string[]>(names: KT) {
        type K = KT[keyof KT & number];
        return names.reduce<Record<K, LogLevel>>(
            (accum, name: K, idx) => {
                accum[name] = {
                    name: name,
                    value: idx,
                };
                return accum;
            },
            {} as Record<K, LogLevel>,
        );
    })(
        // Log level names in order of importance
        ["Debug", "Info", "Warning", "Error"] as const,
    );

    /**
     * Log levels as LogLevel[]
     */
    readonly log_levels = Object.values(this.log_levels_map);
    private readonly default_log_level = this.log_levels_map.Info;

    private log_group_key_counter = 0;
    private readonly _log_groups = list_property<LogGroup>();
    private readonly default_log_group: LogGroup = this.add_log_group("Default");

    private readonly _current_quest_filename = property<string | undefined>(undefined);
    private readonly _current_quest = property<QuestModel | undefined>(undefined);
    private readonly _current_area = property<AreaModel | undefined>(undefined);
    private readonly _selected_entity = property<QuestEntityModel | undefined>(undefined);
    private readonly _log_messages = list_property<LogMessage>();
    private readonly _log_level = property<LogLevel>(this.default_log_level);
    private readonly _log_group = property<LogGroup>(this.default_log_group);
    private readonly _show_all_log_groups = property<boolean>(true);

    readonly debug: WritableProperty<boolean> = property(false);
    readonly undo = new UndoStack();
    readonly current_quest_filename: Property<string | undefined> = this._current_quest_filename;
    readonly current_quest: Property<QuestModel | undefined> = this._current_quest;
    readonly current_area: Property<AreaModel | undefined> = this._current_area;
    readonly selected_entity: Property<QuestEntityModel | undefined> = this._selected_entity;
    readonly log_messages: ListProperty<LogMessage>;
    readonly log_level: Property<LogLevel> = this._log_level;
    readonly log_group: Property<LogGroup> = this._log_group;
    readonly log_groups: ListProperty<LogGroup> = this._log_groups;

    constructor() {
        this.disposer.add_all(
            gui_store.tool.observe(
                ({ value: tool }) => {
                    if (tool === GuiTool.QuestEditor) {
                        this.undo.make_current();
                    }
                },
                { call_now: true },
            ),

            this.current_quest
                .flat_map(quest => (quest ? quest.npcs : property([])))
                .observe(({ value: npcs }) => {
                    const selected = this.selected_entity.val;

                    if (selected instanceof QuestNpcModel && !npcs.includes(selected)) {
                        this.set_selected_entity(undefined);
                    }
                }),

            this.current_quest
                .flat_map(quest => (quest ? quest.objects : property([])))
                .observe(({ value: objects }) => {
                    const selected = this.selected_entity.val;

                    if (selected instanceof QuestObjectModel && !objects.includes(selected)) {
                        this.set_selected_entity(undefined);
                    }
                }),

            // force refresh
            this._log_level.observe(() => (this._log_messages.val = this._log_messages.val)),
            this._log_group.observe(() => (this._log_messages.val = this._log_messages.val)),
            this._show_all_log_groups.observe(
                () => (this._log_messages.val = this._log_messages.val),
            ),
        );

        this.log_messages = this._log_messages.filtered(this.log_message_predicate);
    }

    dispose(): void {
        this.disposer.dispose();
    }

    set_current_area = (area?: AreaModel): void => {
        this._selected_entity.val = undefined;

        this._current_area.val = area;
    };

    set_selected_entity = (entity?: QuestEntityModel): void => {
        if (entity && this.current_quest.val) {
            this._current_area.val = area_store.get_area(
                this.current_quest.val.episode,
                entity.area_id,
            );
        }

        this._selected_entity.val = entity;
    };

    new_quest = (episode: Episode): void => {
        this.set_quest(create_new_quest(episode));
    };

    // TODO: notify user of problems.
    open_file = async (file: File): Promise<void> => {
        try {
            const buffer = await read_file(file);
            const quest = parse_quest(new ArrayBufferCursor(buffer, Endianness.Little));
            this.set_quest(quest && convert_quest_to_model(quest), file.name);
        } catch (e) {
            logger.error("Couldn't read file.", e);
        }
    };

    save_as = (): void => {
        const quest = this.current_quest.val;
        if (!quest) return;

        let default_file_name = this.current_quest_filename.val;

        if (default_file_name) {
            const ext_start = default_file_name.lastIndexOf(".");
            if (ext_start !== -1) default_file_name = default_file_name.slice(0, ext_start);
        }

        let file_name = prompt("File name:", default_file_name);
        if (!file_name) return;

        const buffer = write_quest_qst(convert_quest_from_model(quest), file_name);

        if (!file_name.endsWith(".qst")) {
            file_name += ".qst";
        }

        const a = document.createElement("a");
        a.href = URL.createObjectURL(new Blob([buffer], { type: "application/octet-stream" }));
        a.download = file_name;
        document.body.appendChild(a);
        a.click();
        URL.revokeObjectURL(a.href);
        document.body.removeChild(a);
    };

    id_changed = (event: PropertyChangeEvent<number>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditIdAction(this.current_quest.val, event)).redo();
        }
    };

    name_changed = (event: PropertyChangeEvent<string>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditNameAction(this.current_quest.val, event)).redo();
        }
    };

    short_description_changed = (event: PropertyChangeEvent<string>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditShortDescriptionAction(this.current_quest.val, event)).redo();
        }
    };

    long_description_changed = (event: PropertyChangeEvent<string>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditLongDescriptionAction(this.current_quest.val, event)).redo();
        }
    };

    translate_entity = (
        entity: QuestEntityModel,
        old_section: SectionModel | undefined,
        new_section: SectionModel | undefined,
        old_position: Vector3,
        new_position: Vector3,
        world: boolean,
    ): void => {
        this.undo
            .push(
                new TranslateEntityAction(
                    entity,
                    old_section,
                    new_section,
                    old_position,
                    new_position,
                    world,
                ),
            )
            .redo();
    };

    rotate_entity = (
        entity: QuestEntityModel,
        old_rotation: Euler,
        new_rotation: Euler,
        world: boolean,
    ): void => {
        this.undo.push(new RotateEntityAction(entity, old_rotation, new_rotation, world)).redo();
    };

    push_create_entity_action = (entity: QuestEntityModel): void => {
        this.undo.push(new CreateEntityAction(entity));
    };

    remove_entity = (entity: QuestEntityModel): void => {
        this.undo.push(new RemoveEntityAction(entity)).redo();
    };

    private async set_quest(quest?: QuestModel, filename?: string): Promise<void> {
        this.undo.reset();

        this._current_area.val = undefined;
        this._selected_entity.val = undefined;

        this._current_quest_filename.val = filename;
        this._current_quest.val = quest;

        if (quest) {
            this._current_area.val = area_store.get_area(quest.episode, 0);

            // Load section data.
            for (const variant of quest.area_variants.val) {
                const sections = await area_store.get_area_sections(quest.episode, variant);
                variant.set_sections(sections);

                for (const object of quest.objects.val.filter(o => o.area_id === variant.area.id)) {
                    try {
                        this.set_section_on_quest_entity(object, sections);
                    } catch (e) {
                        logger.error(e);
                    }
                }

                for (const npc of quest.npcs.val.filter(npc => npc.area_id === variant.area.id)) {
                    try {
                        this.set_section_on_quest_entity(npc, sections);
                    } catch (e) {
                        logger.error(e);
                    }
                }
            }
        } else {
            logger.error("Couldn't parse quest file.");
        }
    }

    private set_section_on_quest_entity = (
        entity: QuestEntityModel,
        sections: SectionModel[],
    ): void => {
        const section = sections.find(s => s.id === entity.section_id.val);

        if (section) {
            entity.set_section(section);
        } else {
            logger.warn(`Section ${entity.section_id.val} not found.`);
        }
    };

    run_current_quest = (): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.quest_runner.run(quest);
        }
    };

    private log_message_predicate = (msg: LogMessage): boolean => {
        return (
            (this._show_all_log_groups.val || msg.log_group.key === this.log_group.val.key) &&
            msg.log_level.value >= this.log_level.val.value
        );
    };

    public set_log_level(level_arg: LogLevel): void {
        const level = this.log_levels.find(lvl => lvl.value === level_arg.value);

        if (!level) {
            throw new Error(`Invalid argument: No such log level "${level_arg.name}"`);
        }

        this._log_level.val = level;
    }

    public set_log_group(group_arg: LogGroup): void {
        const group = this._log_groups.val.find(grp => grp.key === group_arg.key);

        if (!group) {
            throw new Error(`Invalid argument: No such log group "${group_arg.name}"`);
        }

        this._show_all_log_groups.val = false;
        this._log_group.val = group;
    }

    public show_all_log_groups(): void {
        if (!this._show_all_log_groups.val) {
            this._show_all_log_groups.val = true;
        }
    }

    private add_log_group(name: string): LogGroup {
        const new_group: LogGroup = {
            name: name,
            key: this.log_group_key_counter++,
        };
        this._log_groups.push(new_group);
        return new_group;
    }

    private add_log_message(items: any[], level: LogLevel, group: LogGroup): void {
        this._log_messages.push({
            formatted_timestamp: new Date().toISOString(),
            message_contents: items.join(" "),
            log_level: level,
            log_group: group,
        });
    }

    public get_logger(group_name?: string): Logger {
        let group = this.default_log_group;

        // find existing or create new
        if (group_name !== undefined) {
            const found = this.log_groups.val.find(grp => grp.name === group_name);

            if (found) {
                group = found;
            } else {
                group = this.add_log_group(group_name);
            }
        }

        // logger object bound to log group
        return {
            debug: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Debug, group);
            },
            info: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Info, group);
            },
            warning: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Warning, group);
            },
            error: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Error, group);
            },
        };
    }
}

export const quest_editor_store = new QuestEditorStore();
