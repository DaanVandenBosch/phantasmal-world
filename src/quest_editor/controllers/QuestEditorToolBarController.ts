import { GuiStore, GuiTool } from "../../core/stores/GuiStore";
import { AreaStore } from "../stores/AreaStore";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { AreaModel } from "../model/AreaModel";
import { list_property, map } from "../../core/observable";
import { Property } from "../../core/observable/property/Property";
import { undo_manager } from "../../core/undo/UndoManager";
import { Controller } from "../../core/controllers/Controller";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import { create_new_quest } from "../stores/quest_creation";
import { read_file } from "../../core/read_file";
import {
    parse_bin_dat_to_quest,
    parse_qst_to_quest,
    Quest,
    write_quest_qst,
} from "../../core/data_formats/parsing/quest";
import { ArrayBufferCursor } from "../../core/data_formats/cursor/ArrayBufferCursor";
import { Endianness } from "../../core/data_formats/Endianness";
import { convert_quest_from_model, convert_quest_to_model } from "../stores/model_conversion";
import { LogManager } from "../../core/Logger";
import { input } from "../../core/gui/dom";
import { basename } from "../../core/util";

const logger = LogManager.get("quest_editor/controllers/QuestEditorToolBarController");

export type AreaAndLabel = { readonly area: AreaModel; readonly label: string };

export class QuestEditorToolBarController extends Controller {
    private quest_filename?: string;

    readonly vm_feature_active: boolean;
    readonly areas: Property<readonly AreaAndLabel[]>;
    readonly current_area: Property<AreaAndLabel>;
    readonly can_save: Property<boolean>;
    readonly can_undo: Property<boolean>;
    readonly can_redo: Property<boolean>;
    readonly can_select_area: Property<boolean>;
    readonly can_debug: Property<boolean>;
    readonly can_step: Property<boolean>;
    readonly can_stop: Property<boolean>;

    constructor(
        gui_store: GuiStore,
        private readonly area_store: AreaStore,
        private readonly quest_editor_store: QuestEditorStore,
    ) {
        super();

        this.vm_feature_active = gui_store.feature_active("vm");

        // Ensure the areas list is updated when entities are added or removed (the count in the
        // label should update).
        this.areas = quest_editor_store.current_quest.flat_map(quest => {
            if (quest) {
                return quest?.entities_per_area.flat_map(entities_per_area => {
                    return list_property<AreaAndLabel>(
                        undefined,
                        ...area_store.get_areas_for_episode(quest.episode).map(area => {
                            const entity_count = entities_per_area.get(area.id);
                            return {
                                area,
                                label: area.name + (entity_count ? ` (${entity_count})` : ""),
                            };
                        }),
                    );
                });
            } else {
                return list_property<AreaAndLabel>();
            }
        });

        this.current_area = map(
            (areas, area) => areas.find(al => al.area == area)!,
            this.areas,
            quest_editor_store.current_area,
        );

        const quest_loaded = quest_editor_store.current_quest.map(q => q != undefined);
        this.can_save = quest_loaded;
        this.can_select_area = quest_loaded;
        this.can_debug = quest_loaded;

        this.can_undo = map(
            (c, r) => c && !r,
            undo_manager.can_undo,
            quest_editor_store.quest_runner.running,
        );

        this.can_redo = map(
            (c, r) => c && !r,
            undo_manager.can_redo,
            quest_editor_store.quest_runner.running,
        );

        this.can_step = quest_editor_store.quest_runner.paused;

        this.can_stop = quest_editor_store.quest_runner.running;

        this.disposables(
            quest_editor_store.current_quest.observe(() => {
                this.quest_filename = undefined;
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-O", () => {
                const input_element = input();
                input_element.type = "file";
                input_element.multiple = true;
                input_element.onchange = () => {
                    if (input_element.files && input_element.files.length) {
                        this.open_files(Array.prototype.slice.apply(input_element.files));
                    }
                };
                input_element.click();
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Shift-S", this.save_as),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Z", () => {
                undo_manager.undo();
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Shift-Z", () => {
                undo_manager.redo();
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F5", this.debug),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F5", this.stop),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F6", this.resume),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F8", this.step_over),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F7", this.step_in),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F8", this.step_out),
        );
    }

    create_new_quest = async (episode: Episode): Promise<void> =>
        this.quest_editor_store.set_current_quest(create_new_quest(this.area_store, episode));

    // TODO: notify user of problems.
    open_files = async (files: File[]): Promise<void> => {
        try {
            let quest: Quest | undefined;

            const qst = files.find(f => f.name.toLowerCase().endsWith(".qst"));

            if (qst) {
                const buffer = await read_file(qst);
                quest = parse_qst_to_quest(new ArrayBufferCursor(buffer, Endianness.Little));
                this.quest_filename = qst.name;
            } else {
                const bin = files.find(f => f.name.toLowerCase().endsWith(".bin"));
                const dat = files.find(f => f.name.toLowerCase().endsWith(".dat"));

                if (bin && dat) {
                    const bin_buffer = await read_file(bin);
                    const dat_buffer = await read_file(dat);
                    quest = parse_bin_dat_to_quest(
                        new ArrayBufferCursor(bin_buffer, Endianness.Little),
                        new ArrayBufferCursor(dat_buffer, Endianness.Little),
                    );
                    this.quest_filename = bin.name || dat.name;
                }
            }

            if (!quest) {
                logger.error("Couldn't parse quest file.");
            }

            await this.quest_editor_store.set_current_quest(
                quest && convert_quest_to_model(this.area_store, quest),
            );
        } catch (e) {
            logger.error("Couldn't read file.", e);
        }
    };

    set_area = ({ area }: AreaAndLabel): void => {
        this.quest_editor_store.set_current_area(area);
    };

    save_as = (): void => {
        const quest = this.quest_editor_store.current_quest.val;
        if (!quest) return;

        const default_file_name = this.quest_filename && basename(this.quest_filename);

        let file_name = prompt("File name:", default_file_name);
        if (!file_name) return;

        const buffer = write_quest_qst(convert_quest_from_model(quest), file_name);

        if (!file_name.endsWith(".qst")) {
            file_name += ".qst";
        }

        const a = document.createElement("a");
        a.href = URL.createObjectURL(new Blob([buffer], { type: "application/octet-stream" }));
        a.download = file_name;
        document.body.appendChild(a);
        a.click();
        URL.revokeObjectURL(a.href);
        document.body.removeChild(a);
    };

    debug = (): void => {
        const quest = this.quest_editor_store.current_quest.val;

        if (quest) {
            this.quest_editor_store.quest_runner.run(quest);
        }
    };

    resume = (): void => {
        this.quest_editor_store.quest_runner.resume();
    };

    step_over = (): void => {
        this.quest_editor_store.quest_runner.step_over();
    };

    step_in = (): void => {
        this.quest_editor_store.quest_runner.step_into();
    };

    step_out = (): void => {
        this.quest_editor_store.quest_runner.step_out();
    };

    stop = (): void => {
        this.quest_editor_store.quest_runner.stop();
    };
}
