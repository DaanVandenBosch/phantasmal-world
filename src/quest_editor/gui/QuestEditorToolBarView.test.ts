import { QuestEditorToolBarController } from "../controllers/QuestEditorToolBarController";
import { QuestEditorToolBarView } from "./QuestEditorToolBarView";
import { GuiStore } from "../../core/stores/GuiStore";
import { create_area_store } from "../../../test/src/quest_editor/stores/store_creation";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { with_disposer } from "../../../test/src/core/observables/disposable_helpers";
import { LogStore } from "../stores/LogStore";

test("Renders correctly.", () =>
    with_disposer(disposer => {
        const gui_store = disposer.add(new GuiStore());
        const area_store = create_area_store(disposer);
        const log_store = disposer.add(new LogStore());
        const quest_editor_store = disposer.add(
            new QuestEditorStore(gui_store, area_store, log_store),
        );
        const tool_bar = disposer.add(
            new QuestEditorToolBarView(
                disposer.add(
                    new QuestEditorToolBarController(gui_store, area_store, quest_editor_store),
                ),
            ),
        );

        expect(tool_bar.element).toMatchSnapshot();
    }));
