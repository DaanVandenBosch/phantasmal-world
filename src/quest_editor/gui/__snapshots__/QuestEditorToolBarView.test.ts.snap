// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly. 1`] = `
<div
  class="core_ToolBar"
  style="height: 33px;"
>
  <div
    class="core_DropDown"
  >
    <button
      class="core_Button"
    >
      <span
        class="core_Button_inner"
      >
        <span
          class="core_Button_left"
        >
          <span>
            <span
              class="fas fa-file-medical"
            />
          </span>
        </span>
        <span
          class="core_Button_center"
        >
          New quest
        </span>
        <span
          class="core_Button_right"
        >
          <span>
            <span
              class="fas fa-caret-down"
            />
          </span>
        </span>
      </span>
    </button>
    <div
      class="core_Menu"
      hidden=""
      tabindex="-1"
    >
      <div
        class="core_Menu_inner"
      >
        <div
          data-index="0"
        >
          Episode I
        </div>
      </div>
    </div>
  </div>
  <button
    class="core_Button core_FileButton"
    title="Open a quest file (Ctrl-O)"
  >
    <span
      class="core_Button_inner"
    >
      <span
        class="core_Button_left"
      >
        <span>
          <span
            class="fas fa-file"
          />
        </span>
      </span>
      <span
        class="core_Button_center"
      >
        Open file...
      </span>
    </span>
  </button>
  <button
    class="core_Button disabled"
    disabled=""
    title="Save this quest to new file (Ctrl-Shift-S)"
  >
    <span
      class="core_Button_inner"
    >
      <span
        class="core_Button_left"
      >
        <span>
          <span
            class="fas fa-save"
          />
        </span>
      </span>
      <span
        class="core_Button_center"
      >
        Save as...
      </span>
    </span>
  </button>
  <button
    class="core_Button disabled"
    disabled=""
    title="Nothing to undo (Ctrl-Z)"
  >
    <span
      class="core_Button_inner"
    >
      <span
        class="core_Button_left"
      >
        <span>
          <span
            class="fas fa-undo"
          />
        </span>
      </span>
      <span
        class="core_Button_center"
      >
        Undo
      </span>
    </span>
  </button>
  <button
    class="core_Button disabled"
    disabled=""
    title="Nothing to redo (Ctrl-Shift-Z)"
  >
    <span
      class="core_Button_inner"
    >
      <span
        class="core_Button_left"
      >
        <span>
          <span
            class="fas fa-redo"
          />
        </span>
      </span>
      <span
        class="core_Button_center"
      >
        Redo
      </span>
    </span>
  </button>
  <div
    class="core_Select disabled"
  >
    <button
      class="core_Button disabled"
      disabled=""
    >
      <span
        class="core_Button_inner"
      >
        <span
          class="core_Button_center"
        >
           
        </span>
        <span
          class="core_Button_right"
        >
          <span>
            <span
              class="fas fa-caret-down"
            />
          </span>
        </span>
      </span>
    </button>
    <div
      class="core_Menu"
      hidden=""
      tabindex="-1"
    >
      <div
        class="core_Menu_inner"
      />
    </div>
  </div>
</div>
`;
